core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -187,7 +186,7 @@ protected NodesAndVersions buildAllocationIdBasedNodes(ShardRouting shard, boole
             }
 
             if (nodeShardState.storeException() == null) {
-                if (allocationId == null && nodeShardState.version() != -1) {
+                if (allocationId == null && nodeShardState.legacyVersion() != -1) {
```

Comment:
- restore creates a new one (with no allocation id, a fresh allocation id is added when shard is assigned).
- The allocation id has only been added to ShardStateMetaData in v3.0.0.